test/Core.Test/Services/DeviceServiceTests.cs

```diff
@@ -33,4 +37,61 @@ public async Task DeviceSaveShouldUpdateRevisionDateAndPushRegistration()
         await pushRepo.Received().CreateOrUpdateRegistrationAsync("testtoken", id.ToString(),
             userId.ToString(), "testid", DeviceType.Android);
     }
+
+    /// <summary>
+    /// 
+    /// </summary>
+    [Theory, BitAutoData]
+    public async Task UpdateDevicesTrustAsync_Works(
+        SutProvider<DeviceService> sutProvider,
+        Guid currentUserId,
+        Device deviceOne,
+        Device deviceTwo,
+        Device deviceThree)
+    {
+        deviceOne.Identifier = "current_device";
+
+        sutProvider.GetDependency<IDeviceRepository>()
+            .GetManyByUserIdAsync(currentUserId)
+            .Returns(new List<Device>
+            {
+                deviceOne,
+                deviceTwo,
+                deviceThree,
+            });
+
+        var currentDeviceModel = new DeviceKeysUpdateRequestModel
+        {
+            EncryptedPublicKey = "current_encrypted_public_key",
+            EncryptedUserKey = "current_encrypted_user_key",
+        };
+
+        var alteredDeviceModels = new List<OtherDeviceKeysUpdateRequestModel>
+        {
+            new OtherDeviceKeysUpdateRequestModel
+            {
+                DeviceId = deviceTwo.Id,
+                EncryptedPublicKey = "encrypted_public_key_two",
+                EncryptedUserKey = "encrypted_user_key_two",
+            },
+        };
+
+        await sutProvider.Sut.UpdateDevicesTrustAsync("current_device", currentUserId, currentDeviceModel, alteredDeviceModels);
+
+        await sutProvider.GetDependency<IDeviceRepository>()
+            .Received(1)
+            .UpsertAsync(Arg.Is<Device>(d => d.Id == deviceOne.Id && d.EncryptedPublicKey == "current_encrypted_public_key"));
+
+        await sutProvider.GetDependency<IDeviceRepository>()
+            .Received(1)
+            .UpsertAsync(Arg.Is<Device>(d => d.Id == deviceTwo.Id && d.EncryptedPublicKey == "encrypted_public_key_two"));
+
+        await sutProvider.GetDependency<IDeviceRepository>()
+            .Received(1)
+            .UpsertAsync(Arg.Is<Device>(d => d.Id == deviceThree.Id && d.EncryptedPublicKey == null));
+
+        await sutProvider.GetDependency<IDeviceRepository>()
+            .Received(3)
+            .UpsertAsync(Arg.Any<Device>());
+    }
 }
```

Comment:
I think it might be worth adding a test to verify that the `UpdateDevicesTrustAsync` logic properly skips devices that are not trusted. 

Locally tested this:
```C#
  [Theory, BitAutoData]
    public async Task UpdateDevicesTrustAsync_DoesNotUpdateUntrustedDevices(
        SutProvider<DeviceService> sutProvider,
        Guid currentUserId,
        Device deviceOne,
        Device deviceTwo)
    {
        deviceOne.Identifier = "current_device";
        
        // Make deviceTwo untrusted
        deviceTwo.EncryptedUserKey = string.Empty; 

        sutProvider.GetDependency<IDeviceRepository>()
            .GetManyByUserIdAsync(currentUserId)
            .Returns(new List<Device>
            {
                deviceOne,
                deviceTwo,
            });

        var currentDeviceModel = new DeviceKeysUpdateRequestModel
        {
            EncryptedPublicKey = "current_encrypted_public_key",
            EncryptedUserKey = "current_encrypted_user_key",
        };

        var alteredDeviceModels = new List<OtherDeviceKeysUpdateRequestModel>
        {
            new OtherDeviceKeysUpdateRequestModel
            {
                DeviceId = deviceTwo.Id,
                EncryptedPublicKey = "encrypted_public_key_two",
                EncryptedUserKey = "encrypted_user_key_two",
            },
        };

        await sutProvider.Sut.UpdateDevicesTrustAsync("current_device", currentUserId, currentDeviceModel, alteredDeviceModels);

        // Check that UpsertAsync was called for the trusted device
        await sutProvider.GetDependency<IDeviceRepository>()
            .Received(1)
            .UpsertAsync(Arg.Is<Device>(d => d.Id == deviceOne.Id && d.EncryptedPublicKey == "current_encrypted_public_key" && d.EncryptedUserKey == "current_encrypted_user_key"));

        // Check that UpsertAsync was not called for the untrusted device
        await sutProvider.GetDependency<IDeviceRepository>()
            .DidNotReceive()
            .UpsertAsync(Arg.Is<Device>(d => d.Id == deviceTwo.Id));
    }
```

Reply:
Here is a locally tested test for the scenario where the `currentDeviceIdentifier` does not match any of the devices in the repository:
```C#
[Theory, BitAutoData]
public async Task UpdateDevicesTrustAsync_ThrowsNotFoundException_WhenCurrentDeviceIdentifierDoesNotExist(
    SutProvider<DeviceService> sutProvider,
    Guid currentUserId,
    Device deviceOne,
    Device deviceTwo)
{
    deviceOne.Identifier = "some_other_device";
    deviceTwo.Identifier = "another_device";

    sutProvider.GetDependency<IDeviceRepository>()
        .GetManyByUserIdAsync(currentUserId)
        .Returns(new List<Device>
        {
            deviceOne,
            deviceTwo,
        });

    var currentDeviceModel = new DeviceKeysUpdateRequestModel
    {
        EncryptedPublicKey = "current_encrypted_public_key",
        EncryptedUserKey = "current_encrypted_user_key",
    };

    await Assert.ThrowsAsync<NotFoundException>(() =>
        sutProvider.Sut.UpdateDevicesTrustAsync("current_device", currentUserId, currentDeviceModel, Enumerable.Empty<OtherDeviceKeysUpdateRequestModel>()));
}
```

Reply:
And here is a test for the scenario where the current device is included in the `alteredDevices` which should throw a bad request:

```C#
[Theory, BitAutoData]
public async Task UpdateDevicesTrustAsync_ThrowsBadRequestException_WhenCurrentDeviceIsIncludedInAlteredDevices(
    SutProvider<DeviceService> sutProvider,
    Guid currentUserId,
    Device deviceOne,
    Device deviceTwo)
{
    deviceOne.Identifier = "current_device";

    sutProvider.GetDependency<IDeviceRepository>()
        .GetManyByUserIdAsync(currentUserId)
        .Returns(new List<Device>
        {
            deviceOne,
            deviceTwo,
        });

    var currentDeviceModel = new DeviceKeysUpdateRequestModel
    {
        EncryptedPublicKey = "current_encrypted_public_key",
        EncryptedUserKey = "current_encrypted_user_key",
    };

    var alteredDeviceModels = new List<OtherDeviceKeysUpdateRequestModel>
    {
        new OtherDeviceKeysUpdateRequestModel
        {
            DeviceId = deviceOne.Id, // current device is included in alteredDevices
            EncryptedPublicKey = "encrypted_public_key_one",
            EncryptedUserKey = "encrypted_user_key_one",
        },
    };

    await Assert.ThrowsAsync<BadRequestException>(() =>
        sutProvider.Sut.UpdateDevicesTrustAsync("current_device", currentUserId, currentDeviceModel, alteredDeviceModels));
}
```

Reply:
Added all these tests and stories for what they test. Thanks for them!